File: clients/service_provider/lib/models/booking.dart
class Booking {
  final String id;
  final String serviceId;
  final String providerId;
  final String customerId;
  final String customerName;
  final String customerPhone;
  final String customerEmail;
  final String? customerLocation;
  final DateTime scheduledDate;
  final String scheduledTime;
  final BookingStatus status;
  final String? notes;
  final String? customerNotes;
  final double totalAmount;
  final PaymentStatus paymentStatus;
  final String? paymentMethod;
  final String? paymentReference;
  final List<BookingItem> items;
  final BookingLocation? location;
  final List<String> attachments;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? completedAt;
  final DateTime? cancelledAt;
  final String? cancellationReason;

  Booking({
    required this.id,
    required this.serviceId,
    required this.providerId,
    required this.customerId,
    required this.customerName,
    required this.customerPhone,
    required this.customerEmail,
    this.customerLocation,
    required this.scheduledDate,
    required this.scheduledTime,
    required this.status,
    this.notes,
    this.customerNotes,
    required this.totalAmount,
    required this.paymentStatus,
    this.paymentMethod,
    this.paymentReference,
    required this.items,
    this.location,
    required this.attachments,
    required this.createdAt,
    required this.updatedAt,
    this.completedAt,
    this.cancelledAt,
    this.cancellationReason,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'],
      serviceId: json['serviceId'],
      providerId: json['providerId'],
      customerId: json['customerId'],
      customerName: json['customerName'],
      customerPhone: json['customerPhone'],
      customerEmail: json['customerEmail'],
      customerLocation: json['customerLocation'],
      scheduledDate: DateTime.parse(json['scheduledDate']),
      scheduledTime: json['scheduledTime'],
      status: BookingStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
      ),
      notes: json['notes'],
      customerNotes: json['customerNotes'],
      totalAmount: (json['totalAmount'] ?? 0.0).toDouble(),
      paymentStatus: PaymentStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['paymentStatus'],
      ),
      paymentMethod: json['paymentMethod'],
      paymentReference: json['paymentReference'],
      items: (json['items'] as List?)
          ?.map((item) => BookingItem.fromJson(item))
          .toList() ?? [],
      location: json['location'] != null 
          ? BookingLocation.fromJson(json['location']) 
          : null,
      attachments: List<String>.from(json['attachments'] ?? []),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      completedAt: json['completedAt'] != null 
          ? DateTime.parse(json['completedAt']) 
          : null,
      cancelledAt: json['cancelledAt'] != null 
          ? DateTime.parse(json['cancelledAt']) 
          : null,
      cancellationReason: json['cancellationReason'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'serviceId': serviceId,
      'providerId': providerId,
      'customerId': customerId,
      'customerName': customerName,
      'customerPhone': customerPhone,
      'customerEmail': customerEmail,
      'customerLocation': customerLocation,
      'scheduledDate': scheduledDate.toIso8601String(),
      'scheduledTime': scheduledTime,
      'status': status.toString().split('.').last,
      'notes': notes,
      'customerNotes': customerNotes,
      'totalAmount': totalAmount,
      'paymentStatus': paymentStatus.toString().split('.').last,
      'paymentMethod': paymentMethod,
      'paymentReference': paymentReference,
      'items': items.map((item) => item.toJson()).toList(),
      'location': location?.toJson(),
      'attachments': attachments,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'completedAt': completedAt?.toIso8601String(),
      'cancelledAt': cancelledAt?.toIso8601String(),
      'cancellationReason': cancellationReason,
    };
  }

  Booking copyWith({
    String? id,
    String? serviceId,
    String? providerId,
    String? customerId,
    String? customerName,
    String? customerPhone,
    String? customerEmail,
    String? customerLocation,
    DateTime? scheduledDate,
    String? scheduledTime,
    BookingStatus? status,
    String? notes,
    String? customerNotes,
    double? totalAmount,
    PaymentStatus? paymentStatus,
    String? paymentMethod,
    String? paymentReference,
    List<BookingItem>? items,
    BookingLocation? location,
    List<String>? attachments,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? completedAt,
    DateTime? cancelledAt,
    String? cancellationReason,
  }) {
    return Booking(
      id: id ?? this.id,
      serviceId: serviceId ?? this.serviceId,
      providerId: providerId ?? this.providerId,
      customerId: customerId ?? this.customerId,
      customerName: customerName ?? this.customerName,
      customerPhone: customerPhone ?? this.customerPhone,
      customerEmail: customerEmail ?? this.customerEmail,
      customerLocation: customerLocation ?? this.customerLocation,
      scheduledDate: scheduledDate ?? this.scheduledDate,
      scheduledTime: scheduledTime ?? this.scheduledTime,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      customerNotes: customerNotes ?? this.customerNotes,
      totalAmount: totalAmount ?? this.totalAmount,
      paymentStatus: paymentStatus ?? this.paymentStatus,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paymentReference: paymentReference ?? this.paymentReference,
      items: items ?? this.items,
      location: location ?? this.location,
      attachments: attachments ?? this.attachments,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      completedAt: completedAt ?? this.completedAt,
      cancelledAt: cancelledAt ?? this.cancelledAt,
      cancellationReason: cancellationReason ?? this.cancellationReason,
    );
  }
}

enum BookingStatus {
  pending,
  confirmed,
  inProgress,
  completed,
  cancelled,
  noShow,
  rescheduled,
}

extension BookingStatusExtension on BookingStatus {
  String get displayName {
    switch (this) {
      case BookingStatus.pending:
        return 'Pending';
      case BookingStatus.confirmed:
        return 'Confirmed';
      case BookingStatus.inProgress:
        return 'In Progress';
      case BookingStatus.completed:
        return 'Completed';
      case BookingStatus.cancelled:
        return 'Cancelled';
      case BookingStatus.noShow:
        return 'No Show';
      case BookingStatus.rescheduled:
        return 'Rescheduled';
    }
  }

  String get color {
    switch (this) {
      case BookingStatus.pending:
        return 'orange';
      case BookingStatus.confirmed:
        return 'blue';
      case BookingStatus.inProgress:
        return 'purple';
      case BookingStatus.completed:
        return 'green';
      case BookingStatus.cancelled:
        return 'red';
      case BookingStatus.noShow:
        return 'grey';
      case BookingStatus.rescheduled:
        return 'amber';
    }
  }
}

enum PaymentStatus {
  pending,
  paid,
  partiallyPaid,
  refunded,
  failed,
}

extension PaymentStatusExtension on PaymentStatus {
  String get displayName {
    switch (this) {
      case PaymentStatus.pending:
        return 'Pending';
      case PaymentStatus.paid:
        return 'Paid';
      case PaymentStatus.partiallyPaid:
        return 'Partially Paid';
      case PaymentStatus.refunded:
        return 'Refunded';
      case PaymentStatus.failed:
        return 'Failed';
    }
  }
}

class BookingItem {
  final String id;
  final String name;
  final String? description;
  final int quantity;
  final double unitPrice;
  final double totalPrice;

  BookingItem({
    required this.id,
    required this.name,
    this.description,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
  });

  factory BookingItem.fromJson(Map<String, dynamic> json) {
    return BookingItem(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      quantity: json['quantity'],
      unitPrice: (json['unitPrice'] ?? 0.0).toDouble(),
      totalPrice: (json['totalPrice'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'quantity': quantity,
      'unitPrice': unitPrice,
      'totalPrice': totalPrice,
    };
  }
}

class BookingLocation {
  final double latitude;
  final double longitude;
  final String address;
  final String? landmark;
  final String? directions;

  BookingLocation({
    required this.latitude,
    required this.longitude,
    required this.address,
    this.landmark,
    this.directions,
  });

  factory BookingLocation.fromJson(Map<String, dynamic> json) {
    return BookingLocation(
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      address: json['address'] ?? '',
      landmark: json['landmark'],
      directions: json['directions'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'latitude': latitude,
      'longitude': longitude,
      'address': address,
      'landmark': landmark,
      'directions': directions,
    };
  }
}

File: services/booking-service/src/models/Booking.js
const mongoose = require('mongoose');

const BookingItemSchema = new mongoose.Schema({
  serviceId: {
    type: String,
    required: true,
  },
  serviceName: {
    type: String,
    required: true,
  },
  quantity: {
    type: Number,
    required: true,
    min: 1,
  },
  unitPrice: {
    type: Number,
    required: true,
    min: 0,
  },
  totalPrice: {
    type: Number,
    required: true,
    min: 0,
  },
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {},
  },
});

const BookingLocationSchema = new mongoose.Schema({
  type: {
    type: String,
    enum: ['Point'],
    required: true,
  },
  coordinates: {
    type: [Number],
    required: true,
    validate: {
      validator: function(v) {
        return v.length === 2;
      },
      message: 'Coordinates must be [longitude, latitude]',
    },
  },
  address: {
    type: String,
    required: true,
  },
  county: String,
  subCounty: String,
  ward: String,
  landmark: String,
});

const BookingSchema = new mongoose.Schema({
  // Core identifiers
  bookingId: {
    type: String,
    required: true,
    unique: true,
    index: true,
  },
  
  // Participants
  farmerId: {
    type: String,
    required: true,
    index: true,
  },
  providerId: {
    type: String,
    index: true,
  },
  
  // Farmer details
  farmerName: {
    type: String,
    required: true,
  },
  farmerPhone: {
    type: String,
    required: true,
  },
  farmerEmail: String,
  
  // Provider details (populated when matched)
  providerName: String,
  providerPhone: String,
  providerEmail: String,
  
  // Booking details
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: true,
  },
  category: {
    type: String,
    required: true,
    enum: ['veterinary', 'crop_advisory', 'equipment_rental', 'labor', 'transport', 'other'],
  },
  
  // Scheduling
  requestedDate: {
    type: Date,
    required: true,
  },
  requestedTime: String,
  scheduledDate: Date,
  scheduledTime: String,
  estimatedDuration: {
    type: Number, // in minutes
    required: true,
  },
  
  // Location
  location: {
    type: BookingLocationSchema,
    required: true,
  },
  
  // Services/Items
  items: [BookingItemSchema],
  
  // Pricing
  totalAmount: {
    type: Number,
    required: true,
    min: 0,
  },
  currency: {
    type: String,
    default: 'KES',
  },
  
  // Status tracking
  status: {
    type: String,
    required: true,
    enum: [
      'pending',           // Just created, waiting for provider match
      'matched',           // Provider found, waiting for acceptance
      'accepted',          // Provider accepted, payment pending
      'confirmed',         // Payment completed, booking confirmed
      'in_progress',       // Service is being delivered
      'completed',         // Service completed, waiting for confirmation
      'cancelled_by_farmer',
      'cancelled_by_provider',
      'disputed',
      'resolved',
    ],
    default: 'pending',
    index: true,
  },
  
  // Payment tracking
  paymentStatus: {
    type: String,
    enum: ['pending', 'processing', 'completed', 'failed', 'refunded'],
    default: 'pending',
    index: true,
  },
  paymentMethod: String,
  paymentReference: String,
  escrowId: String,
  
  // Communication
  notes: String,
  farmerNotes: String,
  providerNotes: String,
  
  // Attachments
  attachments: [{
    type: String,
    url: String,
    filename: String,
    uploadedAt: Date,
  }],
  
  // Ratings & Reviews
  farmerRating: {
    type: Number,
    min: 1,
    max: 5,
  },
  providerRating: {
    type: Number,
    min: 1,
    max: 5,
  },
  farmerReview: String,
  providerReview: String,
  
  // Timestamps
  createdAt: {
    type: Date,
    default: Date.now,
    index: true,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
  matchedAt: Date,
  acceptedAt: Date,
  confirmedAt: Date,
  startedAt: Date,
  completedAt: Date,
  cancelledAt: Date,
  
  // Cancellation details
  cancellationReason: String,
  cancellationNotes: String,
  cancelledBy: String,
  
  // Metadata
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {},
  },
  
  // Tracking
  matchingAttempts: {
    type: Number,
    default: 0,
  },
  lastMatchingAttempt: Date,
  
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
});

// Indexes for performance
BookingSchema.index({ farmerId: 1, status: 1 });
BookingSchema.index({ providerId: 1, status: 1 });
BookingSchema.index({ status: 1, createdAt: -1 });
BookingSchema.index({ location: '2dsphere' });
BookingSchema.index({ category: 1, status: 1 });
BookingSchema.index({ requestedDate: 1, status: 1 });

// Virtual for total duration
BookingSchema.virtual('actualDuration').get(function() {
  if (this.startedAt && this.completedAt) {
    return Math.round((this.completedAt - this.startedAt) / (1000 * 60)); // in minutes
  }
  return null;
});

// Pre-save middleware
BookingSchema.pre('save', function(next) {
  this.updatedAt = new Date();
  next();
});

// Methods
BookingSchema.methods.canBeCancelled = function() {
  return ['pending', 'matched', 'accepted', 'confirmed'].includes(this.status);
};

BookingSchema.methods.canBeRescheduled = function() {
  return ['confirmed'].includes(this.status);
};

BookingSchema.methods.isActive = function() {
  return ['confirmed', 'in_progress'].includes(this.status);
};

BookingSchema.methods.isCompleted = function() {
  return ['completed', 'resolved'].includes(this.status);
};

BookingSchema.methods.isCancelled = function() {
  return this.status.startsWith('cancelled_');
};

// Static methods
BookingSchema.statics.findByFarmer = function(farmerId, options = {}) {
  const query = { farmerId };
  
  if (options.status) {
    query.status = options.status;
  }
  
  if (options.dateRange) {
    query.createdAt = {
      $gte: options.dateRange.start,
      $lte: options.dateRange.end,
    };
  }
  
  return this.find(query)
    .sort({ createdAt: -1 })
    .limit(options.limit || 50);
};

BookingSchema.statics.findByProvider = function(providerId, options = {}) {
  const query = { providerId };
  
  if (options.status) {
    query.status = options.status;
  }
  
  if (options.dateRange) {
    query.createdAt = {
      $gte: options.dateRange.start,
      $lte: options.dateRange.end,
    };
  }
  
  return this.find(query)
    .sort({ createdAt: -1 })
    .limit(options.limit || 50);
};

BookingSchema.statics.findPendingMatches = function(options = {}) {
  const query = { 
    status: 'pending',
    matchingAttempts: { $lt: options.maxAttempts || 5 },
  };
  
  if (options.olderThan) {
    query.createdAt = { $lt: options.olderThan };
  }
  
  return this.find(query).sort({ createdAt: 1 });
};

module.exports = mongoose.model('Booking', BookingSchema);
